lib/Conversion/HWToLLVM/HWToLLVM.cpp

```diff
@@ -367,6 +418,26 @@ struct HWArrayCreateOpConversion
           op->getLoc(), arrayTy, arr, castInput, rewriter.getI32ArrayAttr(i));
     }
 
+    return arr;
+  }
+
+  LogicalResult
+  matchAndRewrite(hw::ArrayCreateOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    auto arrayTy = typeConverter->convertType(op->getResult(0).getType());
+    if (!arrayTy) {
+      op.emitOpError("cannot convert result type");
+      return failure();
+    }
+
+    Value arr;
+    if (isConst(op)) {
+      arr = convertConstArray(op, adaptor, rewriter, arrayTy);
+    } else {
+      arr = convertDynamicArray(op, adaptor, rewriter, arrayTy);
+    }
```

Comment:
Instead of dispatching here, should we maybe implement two separate conversion patterns for constant and non-constant `array_create`? Maybe that would be more modular? And have `isConst` in the `match` function and the content of the two helper functions in the `rewrite` of the corresponding pattern.

Reply:
Is there an example somewhere of how to use match and rewrite separately? In all the conversions I've looked at it uses `matchAndRewrite`, so I'm not exactly sure what the function signatures should look like.

Reply:
I have some commit you could take a look at [here](https://github.com/maerhart/circt/commit/216b90668581c15bfd447d1e35077e1212a7c8de)

Reply:
I made an attempt here: https://github.com/zyedidia/circt/commit/12a06495d9ab17b2850a0202cbcd0b8d574117c8, but when I run the tests they fail with messages such as the following:

```
circt/test/Conversion/HWToLLVM/convert_aggregates.mlir:96:8: error: null operand found
  %2 = hw.array_create %0, %1 : i32
       ^
circt/test/Conversion/HWToLLVM/convert_aggregates.mlir:96:8: note: see current operation: %3 = "builtin.unrealized_conversion_cast"(<<NULL VALUE>>) : (<<NULL TYPE>>) -> !hw.array<2xi32>
```

I'm not sure what I'm missing here that is causing this problem.

Reply:
Can you try to insert the load again that you removed at [line 399](https://github.com/zyedidia/circt/commit/12a06495d9ab17b2850a0202cbcd0b8d574117c8#diff-54489aa272c4db27e53e3b13fcc78510bd787eba4d4bd1749a5c7c10d2802bf3L399)? I think that should fix the issue.

Reply:
Ah yes, thanks for catching that! That was silly of me 😅